docs/COMPLEX_COMPOSITION.md

```diff
@@ -18,6 +18,65 @@ Where Guard rule is an umbrella term for clause, query block, `when` block or na
 
 1. Named rule blocks allow for re-usability, improved composition, and reduced verbosity and repetition. This document will focus on demonstrating these features of named rule blocks in-depth.
 
+## Clause evaluation
+
+Clauses are evaluated on the entire query result. Let us look at some examples to understand this better.
+
+```
+let s3_buckets = Resources.*[ Type == "AWS::S3::Bucket" ]
+
+rule check_s3_bucket when %s3_buckets !empty {
+  %s3_buckets.Properties.AccessControl exists               << Clause 1 >>
+  %s3_buckets.Properties.LoggingConfiguration exists        << Clause 2 >>
+}
+```
+
+In the above example `check_s3_bucket` block evaluates to the following,
+- When the input template has S3 buckets,
+  - All S3 buckets should have `AccessControl` property defined **AND** All S3 buckets should have `LoggingConfiguration` defined.
```

Comment:
I think we should change `should` to `must` here

Reply:
We need an introduction to say that each **AND** clause statement in Guard is an independent statement today. Introduce a simple check like 
```
kind == 'POD'
version == /v1/
```
then prefix `kind` and `version` with `%pods`. Continue with the fact that they are still independent clauses that are evaluated for all values of `%pods` one after the other in that order. 
```
%pods.kind == 'POD'
%pods.version == /v1/
```
You can represent this in pseudocode form like 
```
   for each_pod in %pods:
        each_pod.kind == 'POD'

   for each_pod in %pods:
        each_pod.version == /v1/
```

The show a conjoint AND
```
%pods {
     kind == 'POD'
     version == /v1/
}
```
with the pseudocode 
```
for each_pod in %pods:
     kind == 'POD'
     version == /v1/
```

This explains why independent and conjoint **AND** clauses are equivalent outcomes. Then introduce OR 

```
%pods.kind == 'POD' OR
%pods.version == /v1/
```
and expand to pseudocode form as 
```
   for each_pod in %pods:
        each_pod.kind == 'POD'
   OR
   for each_pod in %pods:
        each_pod.version == /v1/
```
vs conjoint 
```
%pods {
       kind == 'POD' or 
       version == /v1/
}
```
in pseudocode form as 
```
for each_pod in %pods:
     each_pod.kind == 'POD' or each_pod.version == /v1/
```

and helps with the understanding. So conjoint forms is the preferable form of expression